docs/sources/experiments/algorithm.md

```diff
@@ -0,0 +1,182 @@
+
+`all`, with a single value of `True`, will include a feature group with all defined features. If no feature group definition is sent, this is the default.
+
+Either way, at the end of this process the experiment will be aware of some list of feature groups, even if the list is just length 1 with all features as one group.
+
+#### Feature Group Mixing
+A few basic feature group mixing strategies are implemented: `leave-one-in`, `leave-one-out`, and `all`. These are sent in the experiment definition as a list, so different strategies can be tried in the same experiment. Each included strategy will be applied to the list of feature groups from the previous step, to convert them into
+
+For instance, 'leave-one-in' will cycle through each feature group, and for each one create a list of features that just represents that feature group, so for some matrices we would only use features from that particular group. `leave-one-out` does the opposite, for each feature group creating a list of features that includes all other feature groups but that one. `all` just creates a list of features that represents all feature groups together.
+
+### Iteration and Matrix Creation
+
+At this point, matrices are created by looping through all train/test splits and data subsets (e.g. feature groups, state definitions), grabbing the data corresponding to each from the database, and assembling that data into a design matrix that is saved along with the metadata that defines it.
+
+As an example, if the experiment defines 3 train/test splits (one test per train in this example, for simplicity), 3 feature groups that have are mixed using the 'leave-one-out' and 'all' strategies, and 1 state definition, we'll expect 18 matrices to be saved: 9 splits after multiplying the time splits by the feature groups, and each one creating a train and test matrix.
+
+#### Retrieving Data and Saving Completed Matrix
+
+How do we get the data for an individual matrix out of the database?
+
+1. Create an entity-date table for this specific matrix. If it is a test matrix, the table is made up of all valid entity dates. These dates come from the entity-date-state table for the experiment, filtered down to the entity-date pairs that match both *the state filter and the list of as-of-dates for this matrix*. If it is a train matrix, the table is made up of all valid *and labeled* entity dates. The same valid filter used in test matrices applies, but it also joins with the labels table for this experiment on the label name, label type, and label timespan to filter out unlabeled examples.
+
+2. Write features data from tables to disk in CSV format using a COPY command, table by table. Each table is joined with the matrix-specific entity-date table to only include the desired rows.
+
+3. Write labels data to disk in CSV format using a COPY command. It is joined with the matrix-specific entity-date table to only include the desired rows.
```

Comment:
Actually skipping disk would require some substantial rearchitecture, not to mention I think we generally want those to stick around post-modeling for inspection, if not caching.

I think that the forthcoming HDF5 support (mostly finished), and to an extent the forthcoming S3 support is a more robust and workable answer for disk space considerations

Reply:
Sure, I was hoping something like HDF5 could be used, but wasn't sure it applied here.

I absolutely think it makes sense to stream/backup results to S3 for later inspection, rather than assume a non-transient HDD; though, an S3 backend is often not performant enough for programmatic reference mid-run.

So, regardless, there might be cases where we would want or need some more-pluggable interfaces, allowing for tighter integration, (such as passing a file descriptor or file-like object rather than a hard disk path).

But, yes, just off-topic thinking! Sounds good.